super_editor/example/lib/demos/components/demo_text_with_hint.dart

```diff
@@ -174,23 +176,176 @@ class HeaderWithHintComponentBuilder implements ComponentBuilder {
+      maxWidth: maxWidth,
+      padding: padding,
+      blockType: blockType,
+      text: text,
+      textStyleBuilder: textStyleBuilder,
+      textDirection: textDirection,
+      textAlignment: textAlignment,
+      selection: selection,
+      selectionColor: selectionColor,
+      caret: caret,
+      caretColor: caretColor,
+      highlightWhenEmpty: highlightWhenEmpty,
+      shouldShowHintText: shouldShowHintText,
+    );
+  }
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      super == other &&
+          other is ParagraphWithHintComponentViewModel &&
+          runtimeType == other.runtimeType &&
+          nodeId == other.nodeId &&
+          blockType == other.blockType &&
+          text == other.text &&
+          textStyleBuilder == other.textStyleBuilder &&
+          textDirection == other.textDirection &&
+          textAlignment == other.textAlignment &&
+          selection == other.selection &&
+          selectionColor == other.selectionColor &&
+          caret == other.caret &&
+          caretColor == other.caretColor &&
+          highlightWhenEmpty == other.highlightWhenEmpty &&
+          shouldShowHintText == other.shouldShowHintText;
+
+  @override
+  int get hashCode =>
+      super.hashCode ^
+      nodeId.hashCode ^
+      blockType.hashCode ^
+      text.hashCode ^
+      textStyleBuilder.hashCode ^
+      textDirection.hashCode ^
+      textAlignment.hashCode ^
+      selection.hashCode ^
+      selectionColor.hashCode ^
+      caret.hashCode ^
+      caretColor.hashCode ^
+      highlightWhenEmpty.hashCode ^
+      shouldShowHintText.hashCode;
+}
+
+class ParagraphWithHintStyler extends SingleColumnLayoutStylePhase {
+  ParagraphWithHintStyler(this.composer) {
+    composer.selectionNotifier.addListener(markDirty);
+  }
+
+  final DocumentComposer composer;
+
+  @override
+  void dispose() {
+    composer.selectionNotifier.removeListener(markDirty);
+    super.dispose();
+  }
+
+  @override
+  SingleColumnLayoutViewModel style(
+    Document document,
+    SingleColumnLayoutViewModel viewModel,
+  ) {
+    return SingleColumnLayoutViewModel(
+      padding: viewModel.padding,
+      componentViewModels: [
+        for (final previousViewModel in viewModel.componentViewModels)
+          _applySelection(document, previousViewModel.copy()),
+      ],
+    );
+  }
+
+  SingleColumnLayoutComponentViewModel _applySelection(
+    Document document,
+    SingleColumnLayoutComponentViewModel viewModel,
+  ) {
+    final node = document.getNodeById(viewModel.nodeId)!;
+
+    if (node is TextNode && viewModel is ParagraphComponentViewModel) {
+      final blockAttribution = viewModel.blockType;
+      if (!(const [header1Attribution, header2Attribution, header3Attribution]).contains(blockAttribution)) {
+        return viewModel;
+      }
+
+      final documentSelection = composer.selection;
+      final shouldShowHintText = documentSelection?.isCollapsed == true;
```

Comment:
I tried doing something like
```suggestion
      final shouldShowHintText = documentSelection?.isCollapsed == true && viewModel.caretColor != null;
```
But `viewModel.caretColor` is always null. Any idea why?

Reply:
I can't quite tell where this is in the code because GitHub is just showing me a snippet, but it looks like you're implementing the custom style phase?

If so, first, you don't want to make styling decisions based on the `viewModel` that you're given, because you have no control over what has, or hasn't been done to that `viewModel` up to this point. For example, in this case, the reason `caretColor` is null is because the selection styler runs after the custom style phases. This is because user selection is the most volatile behavior, and I didn't want to re-run all the custom style phases every time the user taps somewhere, or drags across a new character.

Further, I'm guessing that the official policy for "when to show hint text" probably isn't determined by the caret color, right? I think your rule here should be "Show the hint text when the selection is collapsed, within this node". That condition would be:

```
final shouldShowHintText = documentSelection != null && documentSelection!.isCollapsed && documentSelection!.extent.nodeId == node.id;
```

I may have gotten some of those terms wrong from memory, but I think that's what you're looking for, right?

Reply:
`viewModel.caretColor != null`, I meant `viewModel.caret != null` but copied the wrong text... 🤦‍♂️ 

Still, it would be incorrect given the presenter pipeline (selection styler runs after the custom stylers). Thanks for explaining it once again. It's getting there 🙂